### PR TITLE
Xep tool

### PR DESCRIPTION
**Pull request by ludwikbukowski**
Tool for fully-automatic production of available XEPs.

### CODE REVIEW

File: apps/ejabberd/src/ejabberd_c2s.hrl [hunk 1]
```diff
@@ -1,6 +1,6 @@
 -include_lib("ejabberd/include/mod_privacy.hrl").
 -define(GEN_FSM, p1_fsm_old).
-
+-xep([{xep, 198}, {version, "1.3"}]).
```
Comment (studzien) [inline apps/ejabberd/src/ejabberd_c2s.hrl]: Why in the .hrl file while other c2s XEP attributes are defined in the .erl file?

Comment (ludwikbukowski) [inline apps/ejabberd/src/ejabberd_c2s.hrl]: by mistake... thanks @studzien !

**Comment by michalwski** [comment]
@ludwikbukowski could you please add some doc how to use your tool? How about adding and new target to Makefile that will run your script?

File: apps/ejabberd/src/mod_amp.erl [hunk 2]
```diff
@@ -6,7 +6,7 @@
 %% This work was sponsored by Grindr LLC
 
 -behavior(gen_mod).
-
+-xep([{xep, 79}, {vesion, "1.2"}, {comment, "partial implemented."}]).
```
Comment (Dzol) [inline apps/ejabberd/src/mod_amp.erl]: Should be "partially implemented" or "partial implementation".

File: tools/xep_tool/xep_tool.escript [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    F = fun(_, {Name, Url}, {Num, BuildingTable}) ->
+        Add = case Num rem 4 of
+                  0 ->
+                      "\n";
+                  _ ->
+                      " "
+              end,
+        {Num + 1, BuildingTable ++ "[" ++ Name ++ "](" ++ Url ++ ") |" ++ Add}
+    end,
+    {_, TableListElement} = maps:fold(F, {1, ""}, Map),
```
Comment (studzien) [inline tools/xep_tool/xep_tool.escript]: What's the decision behind choosing maps here?

Comment (ludwikbukowski) [inline tools/xep_tool/xep_tool.escript]: I'm sure I can replace it with `dict`.

**Comment by studzien** [comment]
My two (ok, three) cents:
* I've noticed that the script is making a HTTP request for the same XEP several times if it occurs in more than one module. Maybe it would be a good idea to pick unique values first :)
* If proper permission are set for the ``xep_tool.escript`` it might be executed directly, without calling ``escript``. Personally, I find it more convenient.
* Maybe it would be a good idea to print a standard "usage" line for the script if it's called without any arguments (or with too many), so it's self-explanatory.

File: tools/xep_tool/xep_tool.escript (hunk 3 above)
Comment (studzien) [inline tools/xep_tool/xep_tool.escript]: So, what the idea behind ``dict`` then? :)

File: tools/xep_tool/xep_tool.escript [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+-record(xep, {xep :: xep(), name :: name(), url :: url(), version :: ver()}).
+-record(module_xep, {xeps = [], module :: module()}).
+
+-spec get_xep(module()) -> {module(), list({atom(), any()})}.
+get_xep(Module) ->
+    Info = Module:module_info(attributes),
+    Values = [Value || {xep, Value} <- Info],
+    {Module, Values}.
+
+-spec match_zeros(xep()) -> string().
```
Comment (studzien) [inline tools/xep_tool/xep_tool.escript]: after changing the returned value to iolist now specs don't match, perhaps also in some other functions

File: tools/xep_tool/xep_tool.escript [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    do(?FILE_NAME);
+
+main([InDir, Output]) ->
+    code:set_path(code:get_path() ++ [InDir]),
+    do(Output);
+
+main(_) ->
+    io:format("~n ***  Bad Arguments. "
+              "Try:~nxep_tool.escript <path_to_ebin_files>"
+              "[output_file]~n~nExample usage (from mongooseim repo root):~n"
```
Comment (studzien) [inline tools/xep_tool/xep_tool.escript]: This prints:
```
 ***  Bad Arguments. Try:
xep_tool.escript <path_to_ebin_files>[output_file]
```
so a space is missing between arguments

File: apps/ejabberd/src/ejabberd_s2s_in.erl [hunk 6]
```diff
@@ -26,7 +26,8 @@
 
 -module(ejabberd_s2s_in).
 -author('alexey@process-one.net').
-
+-xep([{xep, 175}, {version, "1.2"}]).
+-xep([{xep, 198}, {version, "1.3"}]).
```
Comment (studzien) [inline apps/ejabberd/src/ejabberd_s2s_in.erl]: I think we don't support the Stream Management here

File: apps/ejabberd/src/ejabberd_s2s_out.erl [hunk 7]
```diff
@@ -26,7 +26,8 @@
 
 -module(ejabberd_s2s_out).
 -author('alexey@process-one.net').
-
+-xep([{xep, 175}, {version, "1.2"}]).
+-xep([{xep, 198}, {version, "1.3"}]).
```
Comment (studzien) [inline apps/ejabberd/src/ejabberd_s2s_out.erl]: I think we don't support the Stream Management here

File: apps/ejabberd/src/ejabberd_service.erl [hunk 8]
```diff
@@ -26,7 +26,8 @@
 
 -module(ejabberd_service).
 -author('alexey@process-one.net').
-
+-xep([{xep, 114}, {version, "1.6"}]).
+-xep([{xep, 198}, {version, "1.3"}]).
```
Comment (studzien) [inline apps/ejabberd/src/ejabberd_service.erl]: I think we don't support the Stream Management here

File: apps/ejabberd/src/mod_amp.erl [hunk 9]
```diff
@@ -6,7 +6,7 @@
 %% This work was sponsored by Grindr LLC
 
 -behavior(gen_mod).
-
+-xep([{xep, 79}, {vesion, "1.2"}, {comment, "partially implemented."}]).
```
Comment (studzien) [inline apps/ejabberd/src/mod_amp.erl]: a typo in "vesion"

File: apps/ejabberd/src/mod_offline.erl [hunk 10]
```diff
@@ -27,7 +27,11 @@
 
 -module(mod_offline).
 -author('alexey@process-one.net').
-
+-xep([{xep, 160}, {version, "1.0"}]).
+-xep([{xep, 23}, {version, "1.3"}]).
+-xep([{xep, 22}, {version, "1.4"}]).
+-xep([{xep, 85}, {version, "2.1"}]).
+-xep([{xep, 170}, {version, "1.0"}]).
```
Comment (studzien) [inline apps/ejabberd/src/mod_offline.erl]: I'm afraid we don't support XEP-0170 at all, see #242

File: apps/ejabberd/src/mod_vcard.erl [hunk 11]
```diff
@@ -33,7 +33,10 @@
 
 -module(mod_vcard).
 -author('alexey@process-one.net').
-
+-xep([{xep, 54}, {version, "1.2"}]).
+-xep([{xep, 55}, {version, "1.3"}]).
+-xep([{xep, 68}, {version, "1.2"}]).
+-xep([{xep, 82}, {version, "1.1"}]).
```
Comment (studzien) [inline apps/ejabberd/src/mod_vcard.erl]: How does vcard implement/is related to XEP-0082?

**Comment by studzien** [comment]
Generally speaking I'd rather avoid pushing XEP tags wherever we can.
Especially, I think we should include them only in the main feature's module and leave the backends alone.
I believe that in a perfect world we would only use one XEP for a set of modules related to a feature (maybe except informational XEPs).
What do you think?

File: apps/ejabberd/src/jlib.erl [hunk 12]
```diff
@@ -26,7 +26,19 @@
...
 -module(jlib).
 -author('alexey@process-one.net').
-
+-xep([{xep, 59}, {version, "1.0"}]).
+-xep([{xep, 68}, {version, "1.2"}]).
+-xep([{xep, 18}, {version, "0.2"}]).
+-xep([{xep, 175}, {version, "1.2"}]).
+-xep([{xep, 157}, {version, "1.0"}]).
+-xep([{xep, 212}, {version, "1.0"}]).
+-xep([{xep, 114}, {version, "1.6"}]).
```
Comment (studzien) [inline apps/ejabberd/src/jlib.erl]: I think this should go to `ejabberd_service`

File: apps/ejabberd/src/jlib.erl [hunk 13]
```diff
@@ -26,7 +26,19 @@
 
 -module(jlib).
 -author('alexey@process-one.net').
-
+-xep([{xep, 59}, {version, "1.0"}]).
+-xep([{xep, 68}, {version, "1.2"}]).
+-xep([{xep, 18}, {version, "0.2"}]).
```
Comment (studzien) [inline apps/ejabberd/src/jlib.erl]: I think this should go to `ejabberd_c2s`

File: apps/ejabberd/src/jlib.erl [hunk 14]
```diff
@@ -26,7 +26,19 @@
...
+-xep([{xep, 175}, {version, "1.2"}]).
```
Comment (studzien) [inline apps/ejabberd/src/jlib.erl]: I think this should go to `cyrsasl_anonymous`

File: apps/ejabberd/src/jlib.erl [hunk 15]
```diff
@@ -26,7 +26,19 @@
...
+-xep([{xep, 157}, {version, "1.0"}]).
```
Comment (studzien) [inline apps/ejabberd/src/jlib.erl]: I think this should go to `mod_disco`

File: apps/ejabberd/src/jlib.erl [hunk 16]
```diff
@@ -26,7 +26,19 @@
...
+-xep([{xep, 212}, {version, "1.0"}]).
+-xep([{xep, 114}, {version, "1.6"}]).
+-xep([{xep, 85}, {version, "2.1"}]).
```
Comment (studzien) [inline apps/ejabberd/src/jlib.erl]: That's a client-side extension, the only server-side module that is aware of it is `mod_offline`, but I'm not sure if it should land there or in a "static" list of supported XEPs.

File: apps/ejabberd/src/jlib.erl [hunk 17]
```diff
@@ -26,7 +26,19 @@
...
+-xep([{xep, 126}, {version, "1.1"}]).
```
Comment (studzien) [inline apps/ejabberd/src/jlib.erl]: I think this should go to `mod_privacy`

File: apps/ejabberd/src/jlib.erl [hunk 18]
```diff
@@ -26,7 +26,19 @@
...
+-xep([{xep, 170}, {version, "1.0"}]).
```
Comment (studzien) [inline apps/ejabberd/src/jlib.erl]: I don't think we really support this one atm (see #242)

**Comment by lavrin** [comment]
This PR starts to look like a never-ending story ;) Maybe it's time to merge it if the tool itself is OK and then carry on with annotating the modules with subsequent PRs?

**Pull request merged**